src/plus/integrations/authentication/integrationAuthentication.ts

```diff
@@ -37,107 +40,262 @@ export interface IntegrationAuthenticationSessionDescriptor {
+	protected async getBuiltInExistingSession(
+		_?: IntegrationAuthenticationSessionDescriptor,
+	): Promise<ProviderAuthenticationSession | undefined> {
+		return Promise.resolve(undefined);
+	}
+
+	protected override async createSession(
+		descriptor?: IntegrationAuthenticationSessionDescriptor,
+		options?: { authorizeIfNeeded?: boolean },
+	): Promise<ProviderAuthenticationSession | undefined> {
+		const existingSession = await this.getBuiltInExistingSession(descriptor);
+		if (existingSession != null) return existingSession;
+
+		const cloudIntegrations = await this.container.cloudIntegrations;
+		if (cloudIntegrations == null) return undefined;
+
+		let session = await cloudIntegrations.getConnectionSession(this.authProviderId);
+
+		if (session != null && session.expiresIn < 60) {
+			session = await cloudIntegrations.getConnectionSession(this.authProviderId, session.accessToken);
+		}
+
+		if (!session && options?.authorizeIfNeeded) {
+			const authorizeUrl = (await cloudIntegrations.authorize(this.authProviderId))?.url;
+
+			if (!authorizeUrl) return undefined;
+
+			void (await openUrl(authorizeUrl));
+
+			const cancellation = new CancellationTokenSource();
+			const deferredCallback = promisifyDeferred(
+				this.container.uri.onDidReceiveCloudIntegrationAuthenticationUri,
+				this.getUriHandlerDeferredExecutor(),
+			);
 
-			let storedSession: StoredSession | undefined;
 			try {
-				const sessionJSON = await this.container.storage.getSecret(key);
-				if (sessionJSON) {
-					storedSession = JSON.parse(sessionJSON);
-				}
-			} catch (ex) {
-				try {
-					await this.container.storage.deleteSecret(key);
-				} catch {}
-
-				if (!options?.createIfNeeded) {
-					throw ex;
-				}
+				await Promise.race([
+					deferredCallback.promise,
+					this.openCompletionInput(cancellation.token),
+					new Promise<string>((_, reject) =>
+						// eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
+						cancellation.token.onCancellationRequested(() => reject('Cancelled')),
+					),
+					new Promise<string>((_, reject) => setTimeout(reject, 120000, 'Cancelled')),
+				]);
+				session = await cloudIntegrations.getConnectionSession(this.authProviderId);
+			} catch {
+				session = undefined;
+			} finally {
+				cancellation.cancel();
+				cancellation.dispose();
+				deferredCallback.cancel();
 			}
+		}
 
-			if (
-				(options?.createIfNeeded && storedSession == null) ||
-				(storedSession?.expiresAt != null && new Date(storedSession.expiresAt).getTime() < Date.now())
-			) {
-				return this.createSession(providerId, descriptor);
-			}
+		if (!session) return undefined;
 
-			return storedSession as ProviderAuthenticationSession | undefined;
+		return {
+			id: this.getSessionId(descriptor),
+			accessToken: session.accessToken,
+			scopes: descriptor?.scopes ?? [],
+			account: {
+				id: '',
+				label: '',
+			},
+			expiresAt: new Date(session.expiresIn * 1000 + Date.now()),
```

Comment:
Maybe we don't want to include `expiresAt` if the id is github since api is returning 0 (even though it doesn't expire)

Reply:
What is better: to set `undefined` or to set a big value, such as 1 week or 1 year?

Reply:
Actually, that brings up a good point - `expiresAt` on a session right now is the only thing distinguishing cloud token sessions from local ones - perhaps we just set a 1-year `expiresAt` on the tokens for now 👍🏼

Reply:
done
![image](https://github.com/gitkraken/vscode-gitlens/assets/83944326/f429350d-b33d-4eb4-a7ce-0f49bcaf4f1f)